java-frontend/src/main/java/org/sonar/java/regex/RegexParser.java

```diff
@@ -820,14 +824,34 @@ private CharacterClassElementTree parseCharacterRange(CharacterTree startCharact
   }
 
   private PlainCharacterTree plainCharacter(JavaCharacter character) {
+    char c1 = character.getCharacter();
+    if (Character.isHighSurrogate(c1)) {
+      // c1 is in the range from '\uD800' to '\uDBFF', it should be the first char of a series of two,
+      // and it is one 'Supplementary Multilingual Plane' character encoded using UTF-16
+      char c2 = (char) characters.getCurrentChar();
+      if (c2 == '\\') {
+        // skip '\\u'
+        characters.moveNext(2);
+        int codePoint = parseFixedAmountOfHexDigits(4);
+        IndexRange newRange = new IndexRange(character.getRange().getBeginningOffset(), character.getRange().getEndingOffset() + 1);
+        return new PlainCharacterTree(character.getSource(), newRange, Character.toCodePoint(c1, (char) codePoint), true, activeFlags);
+      } else if (Character.isLowSurrogate(c2)) {
+        characters.moveNext();
+        // c2 is in the range from '\uDC00' to '\uDFFF' it's the second part of the UTF-16 code point
+        IndexRange newRange = new IndexRange(character.getRange().getBeginningOffset(), character.getRange().getEndingOffset() + 1);
+        return new PlainCharacterTree(character.getSource(), newRange, Character.toCodePoint(c1, c2), true, activeFlags);
+      } else {
+        LOG.warn("Couldn't parse '{}{}', two high surrogate characters in a row. Please check your encoding.", c1, c2);
+      }
+    }
     return plainCharacter(character, character.getRange());
   }
 
   private PlainCharacterTree plainCharacter(JavaCharacter character, IndexRange range) {
-    return new PlainCharacterTree(source, range, character, activeFlags);
+    return new PlainCharacterTree(source, range, character.getCharacter(), character.isEscapeSequence(), activeFlags);
```

Comment:
I was worried that only checking for multi-unit characters in `plainCharacter(JavaCharacter)` and not `plainCharacter(JavaCharacter, IndexRange)` might lead to breakage, but apparently the latter is only ever called by the former. To me that means it's useless and should just be inlined into `plainCharacter(JavaCharacter)`, so no one accidentally calls it instead of `plainCharacter(JavaCharacter)` in the future.